crypto/bio/bio_mem.c

```diff
@@ -120,34 +134,64 @@ static int mem_free(BIO *bio) {
     return 1;
   }
 
-  BUF_MEM *b = (BUF_MEM *)bio->ptr;
+  BIO_BUF_MEM *bbm = (BIO_BUF_MEM *)bio->ptr;
+  BUF_MEM *b = bbm->buf;
+
   if (bio->flags & BIO_FLAGS_MEM_RDONLY) {
     b->data = NULL;
   }
   BUF_MEM_free(b);
   bio->ptr = NULL;
+
+  OPENSSL_free(bbm);
   return 1;
 }
 
+static void mem_buf_sync(BIO *bio) {
+  if (bio->flags & BIO_FLAGS_MEM_RDONLY) {
+    return;
+  }
+
+  if (bio->init != 0 && bio->ptr != NULL) {
+    BIO_BUF_MEM *bbm = (BIO_BUF_MEM *) bio->ptr;
+    BUF_MEM *b = bbm->buf;
+
+    if (b->data != NULL) {
+      OPENSSL_memmove(b->data, &b->data[bbm->read_off], b->length);
+      bbm->read_off = 0;
+    }
+  }
+}
+
 static int mem_read(BIO *bio, char *out, int outl) {
   BIO_clear_retry_flags(bio);
   if (outl <= 0) {
     return 0;
   }
 
-  BUF_MEM *b = bio->ptr;
+  BIO_BUF_MEM *bbm = (BIO_BUF_MEM *) bio->ptr;
+  BUF_MEM *b = bbm->buf;
+
   int ret = outl;
   if ((size_t)ret > b->length) {
     ret = (int)b->length;
   }
 
   if (ret > 0) {
-    OPENSSL_memcpy(out, b->data, ret);
+    // If the buffer is read-only, |data| is already pointing to the correct
+    // read position. If the buffer is writeable, |data| is pointing to the start
+    // of the buffer, so we need to add |read_off| to get the correct read
+    // position.
```

Comment:
Treating the pointer differently depending on whether the buffer is read-only or not isn't necessary. This adds some complexity to the logic that I think can be removed.